File: jump-game.cpp
//Runtime: 13 ms
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int j = 0;
        for(int i=0;i<nums.size() && i<=j;i++){
            j = max(j, i+nums[i]);
        }
        return j >= (nums.size()-1);
    }
};

File: jump-game.scala
//Runtime: 392 ms
object Solution {
  def canJump(nums: Array[Int]): Boolean = {
    nums.zipWithIndex
      .foldLeft((0, 0))((a, b) =>
        if (a._1 < b._2 + b._1 && b._2 <= a._2) (b._2 + b._1, b._2 + b._1)
        else a)
      ._1 >= (nums.size - 1)
  }
}
